### PR TITLE
feat(hooks): add injection_screening hook to flag untrusted tool output

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

- **New hook**: `injection_screening` (`TOOL_EXECUTE_POST`) screens tool output from web/read/gh/elicit sources for prompt injection patterns
- **Correct signal source**: screens tool *output* (`result_msgs`) not input; requires base.py to collect and pass results to the post hook
- **URL-vs-file discrimination**: the `read` tool is only flagged for URL reads (`http://`/`https://`), not local file reads
- **Pattern coverage**: 11 injection patterns including role-override, instruction-forget, LLM-family delimiters (`<|im_start|>`, `<|system|>`)
- **34 unit tests**: covering detection, false-positive validation, and end-to-end hook behavior

Hardened against the "lethal trifecta" exposure documented in `gptme-contrib/lessons/autonomous/autonomous-operation-safety.md`. Inspired by Claude Code Auto Mode's input-layer protection mechanism.

## Implementation notes

- `base.py`: `result_msgs` is now collected for both Generator and single-return paths, passed to `TOOL_EXECUTE_POST` hooks as a kwarg
- `types.py`: `ToolExecuteHook` protocol unchanged (backward compat); `result_msgs` documented as a kwarg in the docstring
- Existing TOOL_EXECUTE_POST hooks (`token_awareness`, `time_awareness`, `auto_snapshots`, `cwd_changed`, `tool_target_instructions`) updated with `**kwargs` for forward compatibility
- Hook registered in `init_hooks` as a default (non-mode-specific) hook

## Test plan

- [x] `uv run pytest gptme/hooks/tests/test_injection_screening.py -v` — 34 passed
- [x] `uv run mypy gptme/hooks/injection_screening.py gptme/hooks/types.py gptme/tools/base.py` — clean
- [x] `uv run mypy` on all modified hook files — clean

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
Thanks for the Greptile review!

Fixed in e15eedf55:

- **`_is_untrusted_source` branch ordering bug** — restructured the `read` guard so a falsy `tool_content` returns `False` early instead of falling through to `return True`. Previously `read` with `None` or `""` content was incorrectly classified as an untrusted URL read.
- **Added `test_read_with_empty_content_is_trusted`** — covers both `None` and `""` edge cases that were missing from the test suite.

All 35 tests pass.

File: gptme/hooks/injection_screening.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+
+def _is_untrusted_source(tool_name: str, tool_content: str | None) -> bool:
+    """Return True if the tool retrieves untrusted external content."""
+    if tool_name in _UNTRUSTED_SOURCE_TOOLS:
+        # For "read" tool: only flag URL reads, not local file reads
+        if tool_name == "read" and tool_content:
+            return tool_content.strip().startswith(("http://", "https://"))
+        return True
+    return False
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/hooks/injection_screening.py:70]: Fixed in e15eedf55: restructured the guard so `if tool_name == "read"` is handled unconditionally — falsy content returns `False` immediately instead of falling through to `return True`.

File: gptme/hooks/injection_screening.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
+"""Injection screening hook for untrusted tool outputs.
+
+Screens tool outputs from web/email/GitHub sources for prompt injection
+patterns. When detected, prepends an [UNTRUSTED] warning to the model context
+so the model treats the content as untrusted rather than instruction.
+
+Hook type: TOOL_EXECUTE_POST
+"""
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/hooks/injection_screening.py:8]: Fixed in 63ca3a342: docstring now reads "appends" (not "prepends") and clarifies that the warning arrives after the tool output, which is the actual TOOL_EXECUTE_POST behavior.

File: gptme/hooks/injection_screening.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    return False, ""
+
+
+def injection_screening(
+    log=None,
+    workspace=None,
+    tool_use: ToolUse | None = None,
+    result_msgs: list[Message] | None = None,
+    **kwargs,
+) -> Generator[Message, None, None]:
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/hooks/injection_screening.py:92]: Fixed in 63ca3a342: added `log: Log | None`, `workspace: Path | None`, and `**kwargs: Any` type annotations, consistent with all other TOOL_EXECUTE_POST hooks. Added `from ..logmanager import Log`, `from pathlib import Path`, and `from typing import Any` imports.

**Comment by TimeToBuildBob** [comment]
## Greptile findings addressed (63ca3a342)

All three Greptile findings are now fixed:

**P1 — `_is_untrusted_source` branch ordering bug**: `read` with empty/None content now returns `False` immediately instead of falling through to `return True`. Added test `test_read_with_empty_content_is_trusted` covering both `None` and `""` cases.

**P2 — Docstring says "prepends"**: Corrected to "appends" — `TOOL_EXECUTE_POST` runs after `result_msgs` are yielded, so the warning arrives after tool output in the conversation log.

**P2 — Missing type annotations**: Added `log: Log | None`, `workspace: Path | None`, and `**kwargs: Any` to `injection_screening()`, matching every other `TOOL_EXECUTE_POST` hook in the codebase. Added `from ..logmanager import Log`, `from pathlib import Path`, `from typing import Any`.

All 35 tests pass.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by ErikBjare** [comment]
Seems problematic to throw **kwargs on every hook (breaks external hooks/plugins?). Maybe we should pass hooks a single "hook data" object (class could vary by hook type) that we can extend without breaking the API? I think this is how e.g. the `pi` harness does it.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Valid concern — the `**kwargs` approach does mean existing external hooks that match the old `(log, workspace, tool_use)` signature will get a `TypeError` when `base.py` passes `result_msgs=...`, which the registry catches and swallows (so it fails silently rather than loudly).

The HookData pattern is cleaner: `def my_hook(data: ToolExecutePostData)` means adding new fields to the dataclass never touches existing hook signatures. Worth doing.

Both approaches are breaking changes for external hooks (new calling convention either way), so we're not gaining backwards compat — just moving to a better long-term API. Happy to implement HookData in this PR before merging if you'd prefer that over a follow-up. Let me know and I'll get it done.

**Comment by ErikBjare** [comment]
Do the HookData in a follow-up

**Comment by TimeToBuildBob** [comment]
Acknowledged — will track HookData as a follow-up PR. That PR can define typed dataclasses per hook type (`ToolExecutePostData`, etc.) so the calling convention stabilizes before external plugins adopt it.

This PR is ready on my end: CI green, Greptile addressed. Happy to merge once you approve.

**Comment by TimeToBuildBob** [comment]
HookData PR is up: #2656. Typed dataclasses per hook type, no more `**kwargs`, all 437 tests pass.
